[Computed Style Gen] Autogenerate computed style extractor

weinig opened the pull request:
#### e574cf07e45c7955a3f7955240a2d8f479076c34
<pre>
[Computed Style Gen] Autogenerate computed style extractor
<a href="https://bugs.webkit.org/show_bug.cgi?id=292797">https://bugs.webkit.org/show_bug.cgi?id=292797</a>

Reviewed by Darin Adler.

Converts computed style extraction to use autogeneration.

The basic pattern mirrors `Style::Builder`:
- The class `ComputedStyleExtractor` is renamed to `Style::Extractor`.
- `CSSPropertyID` dispatch, that used to be in `ComputedStyleExtractor`
  is now generated in a new class `Style::ExtractorGenerated`.
- Individual property extractions that can be generated are done
  along with `Style::ExtractorGenerated` as static functions in a
  generated struct called `Style::ExtractorFunctions` (mirroring
  `Style::BuilderGenerated` and `Style::BuilderFunctions`).
- Properties that need completely custom extraction are implemented
  as static functions in a `Style::ExtractorCustom` struct (mirroring
  `Style::BuilderCustom`).
- Shared value conversion from style types to CSSValue are implemented
  as static functions in a `Style::ExtractorConverter` struct (mirroring
  `Style::BuilderConverter`).

Generation of the extractors in `Style::ExtractorFunctions` uses new
fields in CSSProperties.json to determine what can be generated.
- `skip-style-extractor` replaces and extends the `computable` field
  telling the generator no extractor should be generated at all (mirroring
  `skip-style-builder`).
- `style-extractor-converter` is used to indicate the suffix for a
    converter in `Style::ExtractorConverter` to use (mirroring
  `style-builder-converter`).
- Where possible, the names of the converters used the builder and
  extractor should be the same. In these cases, instead of specifying
  both `style-builder-converter` and `style-extractor-converter` to
  the same value, and single `style-converter` field can be used.
- `style-extractor-custom` is used to indicate the that a completely
  custom converter is needed (mirroring `style-builder-custom`).
- `shorthand-pattern` is used to indicate which standard shorthand
  pattern should be used. This does not mirror anything in style
  builder, as shorthands are not relevant to building, but has been
  named in such a way that it should be useable for generating
  additional parser functions.

* LayoutTests/fast/css/computed-clip-with-auto-rect-expected.txt:
    - Update results for correct clip() computed value. There was a typo in 776f9cc2
      that made it compute to auto if only the right and bottom sides were auto, but
      really it should have been checking all four sides.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/StyleExtractor.cpp: Added.
* Source/WebCore/style/StyleExtractor.h: Added.
* Source/WebCore/style/StyleExtractorConverter.h: Added.
* Source/WebCore/style/StyleExtractorCustom.h: Added.
* Source/WebCore/style/StyleExtractorGenerated.h: Added..
* Source/WebCore/style/StyleExtractorState.h: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp: Removed.
* Source/WebCore/css/ComputedStyleExtractor.h: Removed.
    - Add/remove files.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
    - Add expectations for moved implementations.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
    - Removed OutlineIsAuto  in favor of matching other RenderStyle properties
      using a standard  `hasAutoOutlineStyle()` function. This allows it to act
      like all the other &quot;auto-functions&quot; for generation.

* Source/WebCore/platform/animation/Animation.h:
    - Adds isNameFilled() which always returns false to make it work
      work with the generator.

* Source/WebCore/css/CSSProperties.json:
    - Add additional fields needed for new generator.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Implement new generator.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Update names where they needed to be tweaked to work with both generators.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
    - Add new tests and update test results for new generated file.

* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/WebAnimation.cpp:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/ShorthandSerializer.h:
* Source/WebCore/css/StyleSheetContents.cpp:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/EditingStyle.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/html/CustomPaintImage.cpp:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderGenerated.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/svg/SVGElement.cpp:
* Source/WebCore/svg/properties/SVGPropertyAnimator.h:
    - Update for renames.

* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
    - Update checker for added/removed fields.

Canonical link: <a href="https://commits.webkit.org/294902@main">https://commits.webkit.org/294902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a8adaf7975f3ac463e780d5a83d9b2d6e4a4ace

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103479 "17 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108655 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54124 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31712 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18230 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/93345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/102955 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53480 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11452 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111033 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30626 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22586 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30987 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89545 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/87273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22212 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32148 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9858 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24973 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30554 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35866 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33685 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31915 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -39,9 +39,9 @@
-#include "ComputedStyleExtractor.h"
 #include "FontSelectionValueInlines.h"
 #include "Quad.h"
+#include "StyleExtractor.h"
```
Can this use a forward declaration instead?

weinig replied:
I don't think so. This is the cpp file and it calls extractor.propertyValue() so it needs the full thing.

darinadler commented on Source/WebCore/platform/animation/Animation.h:
```diff
@@ -222,6 +222,7 @@ class Animation : public RefCounted<Animation> {
     bool isDurationFilled() const { return m_durationFilled; }
     bool isFillModeFilled() const { return m_fillModeFilled; }
     bool isIterationCountFilled() const { return m_iterationCountFilled; }
+    bool isNameFilled() const { return false; } // Needed for property generation generalization.
```
```suggestion
    static bool isNameFilled() { return false; } // Needed for property generation generalization.
```

weinig committed:
[Computed Style Gen] Autogenerate computed style extractor
https://bugs.webkit.org/show_bug.cgi?id=292797

Reviewed by Darin Adler.

Converts computed style extraction to use autogeneration.

The basic pattern mirrors `Style::Builder`:
- The class `ComputedStyleExtractor` is renamed to `Style::Extractor`.
- `CSSPropertyID` dispatch, that used to be in `ComputedStyleExtractor`
  is now generated in a new class `Style::ExtractorGenerated`.
- Individual property extractions that can be generated are done
  along with `Style::ExtractorGenerated` as static functions in a
  generated struct called `Style::ExtractorFunctions` (mirroring
  `Style::BuilderGenerated` and `Style::BuilderFunctions`).
- Properties that need completely custom extraction are implemented
  as static functions in a `Style::ExtractorCustom` struct (mirroring
  `Style::BuilderCustom`).
- Shared value conversion from style types to CSSValue are implemented
  as static functions in a `Style::ExtractorConverter` struct (mirroring
  `Style::BuilderConverter`).

Generation of the extractors in `Style::ExtractorFunctions` uses new
fields in CSSProperties.json to determine what can be generated.
- `skip-style-extractor` replaces and extends the `computable` field
  telling the generator no extractor should be generated at all (mirroring
  `skip-style-builder`).
- `style-extractor-converter` is used to indicate the suffix for a
    converter in `Style::ExtractorConverter` to use (mirroring
  `style-builder-converter`).
- Where possible, the names of the converters used the builder and
  extractor should be the same. In these cases, instead of specifying
  both `style-builder-converter` and `style-extractor-converter` to
  the same value, and single `style-converter` field can be used.
- `style-extractor-custom` is used to indicate the that a completely
  custom converter is needed (mirroring `style-builder-custom`).
- `shorthand-pattern` is used to indicate which standard shorthand
  pattern should be used. This does not mirror anything in style
  builder, as shorthands are not relevant to building, but has been
  named in such a way that it should be useable for generating
  additional parser functions.

* LayoutTests/fast/css/computed-clip-with-auto-rect-expected.txt:
    - Update results for correct clip() computed value. There was a typo in 776f9cc2
      that made it compute to auto if only the right and bottom sides were auto, but
      really it should have been checking all four sides.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/StyleExtractor.cpp: Added.
* Source/WebCore/style/StyleExtractor.h: Added.
* Source/WebCore/style/StyleExtractorConverter.h: Added.
* Source/WebCore/style/StyleExtractorCustom.h: Added.
* Source/WebCore/style/StyleExtractorGenerated.h: Added..
* Source/WebCore/style/StyleExtractorState.h: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp: Removed.
* Source/WebCore/css/ComputedStyleExtractor.h: Removed.
    - Add/remove files.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
    - Add expectations for moved implementations.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
    - Removed OutlineIsAuto  in favor of matching other RenderStyle properties
      using a standard  `hasAutoOutlineStyle()` function. This allows it to act
      like all the other "auto-functions" for generation.

* Source/WebCore/platform/animation/Animation.h:
    - Adds isNameFilled() which always returns false to make it work
      work with the generator.

* Source/WebCore/css/CSSProperties.json:
    - Add additional fields needed for new generator.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Implement new generator.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Update names where they needed to be tweaked to work with both generators.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
    - Add new tests and update test results for new generated file.

* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/WebAnimation.cpp:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/ShorthandSerializer.h:
* Source/WebCore/css/StyleSheetContents.cpp:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/EditingStyle.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/html/CustomPaintImage.cpp:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderGenerated.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/svg/SVGElement.cpp:
* Source/WebCore/svg/properties/SVGPropertyAnimator.h:
    - Update for renames.

* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
    - Update checker for added/removed fields.

Canonical link: https://commits.webkit.org/294902@main
